privacyidea/lib/tokens/pushtoken.py

```
@@ -0,0 +1,578 @@
+        :type param: dict
+
+        :return: nothing
+        """
+        upd_param = {}
+        for k, v in param.items():
+            upd_param[k] = v
+
+        if "serial" in upd_param and "fbtoken" in upd_param and "pubkey" in upd_param:
+            # We are in step 2:
+            if not self.token.rollout_state == "clientwait":
+                raise ParameterError("Invalid state! The token you want to enroll is not in the state 'clientwait'.")
+            if upd_param.get("enrollment_credential") != self.get_tokeninfo("enrollment_credential"):
+                raise ParameterError("Invalid enrollment credential. You are not authorized to finalize this token.")
+            self.del_tokeninfo("enrollment_credential")
+            self.token.rollout_state = "enrolled"
+            self.token.active = True
+            self.add_tokeninfo(PUBLIC_KEY_SMARTPHONE, upd_param.get("pubkey"))
+            self.add_tokeninfo("firebase_token", upd_param.get("fbtoken"))
+            # create a keypair for the server side.
+            from privacyidea.lib.crypto import generate_keypair
+            pub_key, priv_key = generate_keypair(4096)
+            #pub_key = _strip_key(pub_key)
+            #priv_key = _strip_key(priv_key)
+            self.add_tokeninfo(PUBLIC_KEY_SERVER, pub_key)
+            self.add_tokeninfo(PRIVATE_KEY_SERVER, priv_key, "password")
+
+        elif "genkey" in upd_param:
+            # We are in step 1:
+            upd_param["2stepinit"] = 1
+            self.add_tokeninfo("enrollment_credential", geturandom(20, hex=True))
+            # We also store the firebase config, that was used during the enrollment.
+            self.add_tokeninfo(PUSH_ACTION.FIREBASE_CONFIG, param.get(PUSH_ACTION.FIREBASE_CONFIG))
+        else:
+            raise ParameterError("Invalid Parameters. Either provide (genkey) or (serial, fbtoken, pubkey).")
+
+        TokenClass.update(self, upd_param, reset_failcount)
```

If I understand correctly, the ``otpkey`` parameter makes no sense for the push token. However, the underlying ``TokenClass.update`` will process it if it is given, which can have weird effects. Should we maybe throw an error if it is provided?

---

As the otpkey is not used anywhere, I think it would not be necessary.